server/tpm2.go

```diff
@@ -0,0 +1,47 @@
+// SPDX-FileCopyrightText: 2022 ANSSI
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+)
+
+/*
+   Returns @size bytes of random data from the TPM
+*/
+func TPM2_GetRandom(size uint16) ([]byte, error) {
+	rwc, err := tpm2.OpenTPM()
+	if err != nil {
+		return nil, err
+	}
+	defer rwc.Close()
+
+	rbytes, err := tpm2.GetRandom(rwc, size)
+	if err != nil {
+		return nil, err
+	}
+	return rbytes, nil
+}
+
+/*
+	Extends the PCR at the given index with @secret
+	TODO: The following extends the PCR, but does not adds
+	any event log entry. This is sufficient for our needs,
+	but it means that after an ultrablue attestation, the
+	eventlog will differ from the pcrs, thus making any
+	replay fail.
+	NOTE: It seems that it's not possible to add an event log
+	entry directly from the tss stack, and that we need to use
+	exposed UEFI function pointers.
+*/
```

Comment:
Please document the restrictions on `secret` (I guess it must be a certain size that matches the size of the hashes for that particular PCR?).

Reply:
In fact, the only restriction is that the secret isn't `nil`. Even an empty string would work.

Reply:
Ah, because you use PCREvent rather than PCRExtend. Fair enough. But then, is it really fine to just pass the secret as an Event, instead of a more structured binary blob? For instance, what happens the day we want to store this event in the event log, would it make a breaking change to all disks encrypted with ultrablue?

Reply:
@nbouchinet-anssi may have more clues about the format of PCR Events.